File: sdk/typescript/src/contracts/deployer.ts
/**
 * Contract deploy helpers.
 *
 * Responsibilities:
 *  - Build a deploy transaction from code + optional manifest bytes
 *  - Provide a high-level `deploy()` that signs, submits, and waits for receipt
 *  - Small utilities to hash artifacts and extract a contract address from receipt
 */

import type { RpcClient, WaitOpts } from '../tx/send'
import { signSendAndWait } from '../tx/send'
import type { AccessList, UnsignedTx } from '../tx/build'
import { buildDeploy, estimateIntrinsicGas } from '../tx/build'
import type { Signer } from '../wallet/signer'
import { sha3_256 } from '../utils/hash'
import { bytesToHex, hexToBytes } from '../utils/bytes'

export interface DeployParams {
  chainId: number
  from: string
  /** Contract code bytes. If string, accepts hex (0x…) or UTF-8 for dev/demo. */
  code: Uint8Array | string
  /** Optional manifest/ABI bytes packaged with the deploy. */
  manifest?: Uint8Array | string
  /** Optional value to fund contract treasury at deploy. */
  value?: bigint | number | string
  /** Sender nonce and gas price are required; gasLimit can be omitted (estimated). */
  nonce: bigint | number | string
  gasPrice: bigint | number | string
  gasLimit?: bigint | number | string
  accessList?: AccessList
}

/** Build the unsigned deploy transaction (you can sign it with your own wallet flow). */
export function buildDeployTx(p: DeployParams): UnsignedTx {
  const code = normalizeBytes(p.code)
  const manifest = p.manifest ? normalizeBytes(p.manifest) : undefined
  const combinedLen = code.length + (manifest?.length ?? 0)
  const est = estimateIntrinsicGas('deploy', combinedLen, p.accessList)
  const gasLimit = p.gasLimit ?? ((est * 130n) / 100n) // +30% headroom

  return buildDeploy({
    chainId: p.chainId,
    from: p.from,
    code,
    manifest,
    value: p.value,
    nonce: p.nonce,
    gasPrice: p.gasPrice,
    gasLimit,
    accessList: p.accessList
  })
}

export interface DeployResult<Receipt = any> {
  txHash: string
  receipt: Receipt
  /** If the chain returns it in the receipt, this will be populated. */
  contractAddress?: string
  /** Hex (0x…) hashes of the input artifacts for convenience/debugging. */
  codeHashHex: string
  manifestHashHex?: string
}

/**
 * High-level one-shot deploy:
 *  - Builds tx
 *  - Signs with the provided PQ signer
 *  - Submits to RPC
 *  - Waits for receipt
 */
export async function deploy<Receipt = any>(
  client: RpcClient,
  signer: Pick<Signer, 'getPublicKey' | 'sign' | 'alg'>,
  params: DeployParams,
  wait: WaitOpts = {}
): Promise<DeployResult<Receipt>> {
  const tx = buildDeployTx(params)
  const { txHash, receipt } = await signSendAndWait(client, tx, signer, wait)
  const contractAddress = extractContractAddress(receipt)
  const codeBytes = normalizeBytes(params.code)
  const manifestBytes = params.manifest ? normalizeBytes(params.manifest) : undefined
  const codeHashHex = '0x' + bytesToHex(sha3_256(codeBytes))
  const manifestHashHex = manifestBytes ? ('0x' + bytesToHex(sha3_256(manifestBytes))) : undefined

  return { txHash, receipt, contractAddress, codeHashHex, manifestHashHex }
}

/** Attempt to extract a contract address from the receipt (if the node includes it). */
export function extractContractAddress(receipt: any): string | undefined {
  if (receipt && typeof receipt === 'object') {
    if (typeof receipt.contractAddress === 'string') return receipt.contractAddress
    // Some nodes may include it under `address` on deploy receipts:
    if (typeof receipt.address === 'string') return receipt.address
    // Or emit a well-known event; you can extend this if your chain standardizes it.
  }
  return undefined
}

/** Utility: compute 0x-prefixed SHA3-256 of input bytes. */
export function sha3Hex(data: Uint8Array | string): string {
  const b = normalizeBytes(data)
  return '0x' + bytesToHex(sha3_256(b))
}

// ──────────────────────────────────────────────────────────────────────────────
// Internals
// ──────────────────────────────────────────────────────────────────────────────

function normalizeBytes(x: Uint8Array | string): Uint8Array {
  if (x instanceof Uint8Array) return x
  if (typeof x === 'string') {
    if (x.startsWith('0x') || x.startsWith('0X')) return hexToBytes(x)
    return new TextEncoder().encode(x)
  }
  throw new Error('Expected Uint8Array | string')
}

export default {
  buildDeployTx,
  deploy,
  extractContractAddress,
  sha3Hex
}

File: sdk/rust/src/contracts/deployer.rs
//! Contract **package deployer** (manifest + code).
//!
//! This helper bundles a manifest (JSON) and code bytes into a canonical CBOR
//! payload, estimates gas, builds a **Deploy** transaction, signs, submits, and
//! waits for the receipt.
//!
//! ### Deploy payload (CBOR, canonical, integer-keyed map)
//! We encode the deploy package as a 2-field map to keep cross-SDK stability:
//! ```text
//! 0: manifest  (JSON object; validated by the node against spec/manifest.schema.json)
//! 1: code      (bytes; VM-specific, e.g. Python-VM IR bytes)
//! ```
//!
//! This mirrors the Python & TypeScript SDKs.

use crate::error::{Error, Result};
use crate::rpc::http::RpcClient;
use crate::tx::build::{build_deploy, estimate_gas_deploy, GasEstimate};
use crate::tx::send::{build_signed_envelope, send_and_wait};
use crate::types::{AccessListItem, Receipt, Tx};
use crate::utils::hash::sha3_256;
use crate::wallet::signer::TxSigner;
use serde::Serialize;
use serde_cbor::ser::Serializer;
use serde_json::Value as JsonValue;
use std::time::Duration;

/// Options for a deploy (write) transaction.
#[derive(Debug, Clone)]
pub struct DeployOptions<'a, S: TxSigner> {
    pub chain_id: u64,
    pub from: &'a str,         // bech32m "anim1..."
    pub nonce: u64,
    pub gas_price: u64,
    /// If `None`, we'll estimate and add headroom.
    pub gas_limit: Option<u64>,
    /// Value to send with deployment (usually 0).
    pub value: u128,
    /// Optional access-list (builder may construct from manifest hints later).
    pub access_list: &'a [AccessListItem],
    /// PQ signer (Dilithium3 / SPHINCS+).
    pub signer: &'a S,
    /// Poll frequency for receipt.
    pub poll_every: Duration,
    /// Overall timeout for receipt.
    pub timeout: Duration,
}

impl<'a, S: TxSigner> DeployOptions<'a, S> {
    pub fn with_defaults(
        chain_id: u64,
        from: &'a str,
        nonce: u64,
        gas_price: u64,
        signer: &'a S,
    ) -> Self {
        Self {
            chain_id,
            from,
            nonce,
            gas_price,
            gas_limit: None,
            value: 0,
            access_list: &[],
            signer,
            poll_every: Duration::from_secs(1),
            timeout: Duration::from_secs(60),
        }
    }
}

/// Deployer bound to an RPC client.
#[derive(Clone)]
pub struct Deployer<'r> {
    rpc: &'r RpcClient,
}

impl<'r> Deployer<'r> {
    pub fn new(rpc: &'r RpcClient) -> Self {
        Self { rpc }
    }

    /// Encode `{0: manifest, 1: code}` as canonical CBOR bytes.
    pub fn encode_payload(&self, manifest: &JsonValue, code: &[u8]) -> Result<Vec<u8>> {
        // Canonical CBOR with self-describe tag for easy debugging.
        let mut buf = Vec::with_capacity(code.len() + 256);
        let mut ser = Serializer::new(&mut buf);
        ser.self_describe()?;
        ser.canonical();

        // 2-entry map
        ser.serialize_map(Some(2))?;

        // key 0 → manifest (serialize JSON as CBOR via serde)
        ser.serialize_u8(0)?;
        manifest.serialize(&mut ser).map_err(|e| Error::Serde(e.to_string()))?;

        // key 1 → code (bytes)
        ser.serialize_u8(1)?;
        ser.serialize_bytes(code)?;

        ser.end()?;
        Ok(buf)
    }

    /// Compute code hash (SHA3-256) as bytes.
    pub fn code_hash(&self, code: &[u8]) -> [u8; 32] {
        sha3_256(code)
    }

    /// Hex form of code hash (0x-prefixed).
    pub fn code_hash_hex(&self, code: &[u8]) -> String {
        let h = self.code_hash(code);
        let mut s = String::with_capacity(2 + h.len() * 2);
        s.push_str("0x");
        s.push_str(&hex::encode(h));
        s
    }

    /// Heuristic gas estimation for deploy payload + access list.
    pub fn estimate_gas(&self, manifest: &JsonValue, code: &[u8], access_list: &[AccessListItem]) -> Result<GasEstimate> {
        let payload = self.encode_payload(manifest, code)?;
        Ok(estimate_gas_deploy(&payload, access_list))
    }

    /// Build a deploy transaction (unsigned).
    pub fn build_tx(
        &self,
        chain_id: u64,
        from: &str,
        nonce: u64,
        gas_price: u64,
        gas_limit: u64,
        value: u128,
        manifest: &JsonValue,
        code: &[u8],
        access_list: Vec<AccessListItem>,
    ) -> Result<Tx> {
        let payload = self.encode_payload(manifest, code)?;
        Ok(build_deploy(
            chain_id,
            from,
            value,
            nonce,
            gas_price,
            gas_limit,
            payload,
            access_list,
        ))
    }

    /// Deploy a package: encode → (estimate gas) → build → sign → send → await receipt.
    ///
    /// Returns `(tx_hash, receipt)`. The receipt's contract address (if any) is
    /// available via `receipt.contract_address` (per `crate::types`).
    pub async fn deploy_and_wait<S: TxSigner>(
        &self,
        manifest: &JsonValue,
        code: &[u8],
        mut opts: DeployOptions<'_, S>,
    ) -> Result<(String, Receipt)> {
        // Determine gas limit
        let gas_limit = match opts.gas_limit {
            Some(gl) => gl,
            None => self.estimate_gas(manifest, code, opts.access_list)?.suggested_limit,
        };

        // Build tx
        let tx = self.build_tx(
            opts.chain_id,
            opts.from,
            opts.nonce,
            opts.gas_price,
            gas_limit,
            opts.value,
            manifest,
            code,
            opts.access_list.to_vec(),
        )?;

        // Sign quickly to surface encoding/sign issues early
        let _env = build_signed_envelope(&tx, opts.signer)?;

        // Submit & wait
        send_and_wait(self.rpc, &tx, opts.signer, opts.poll_every, opts.timeout).await
    }
}

// ---------------------------------- Tests ------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::AccessListItem;

    #[test]
    fn payload_is_canonical_and_has_two_fields() {
        let rpc = RpcClient::new("http://localhost:0");
        let d = Deployer::new(&rpc);
        let manifest = serde_json::json!({"name":"counter","version":"1.0.0"});
        let code = vec![1u8, 2, 3, 4, 5];
        let bytes = d.encode_payload(&manifest, &code).unwrap();

        // Decode & sanity check
        let val: serde_cbor::Value = serde_cbor::from_slice(&bytes).unwrap();
        match val {
            serde_cbor::Value::Tag(_, inner) => {
                // self-describe tag wraps a map
                match *inner {
                    serde_cbor::Value::Map(m) => {
                        assert_eq!(m.len(), 2);
                    }
                    _ => panic!("expected map inside self-describe tag"),
                }
            }
            _ => panic!("expected self-describe tag"),
        }
    }

    #[test]
    fn code_hash_hex_formats() {
        let rpc = RpcClient::new("http://localhost:0");
        let d = Deployer::new(&rpc);
        let hex = d.code_hash_hex(b"hello");
        assert!(hex.starts_with("0x"));
        assert_eq!(hex.len(), 2 + 64); // 32 bytes → 64 hex chars
    }

    #[test]
    fn estimate_and_build() {
        let rpc = RpcClient::new("http://localhost:0");
        let d = Deployer::new(&rpc);
        let manifest = serde_json::json!({"name":"counter","version":"1.0.0"});
        let code = vec![0u8; 64];
        let gas = d.estimate_gas(&manifest, &code, &[]).unwrap();
        assert!(gas.suggested_limit >= gas.intrinsic);

        let tx = d
            .build_tx(1, "anim1from...", 7, 1_000, gas.suggested_limit, 0, &manifest, &code, vec![])
            .unwrap();
        assert_eq!(tx.chain_id, 1);
        assert_eq!(tx.nonce, 7);
        assert_eq!(tx.gas_price, 1_000);
        assert!(tx.data.len() > 0);
        assert!(tx.to.is_none(), "deploy tx has no 'to' address");
    }
}
